Prevent error when renaming variable

datho7561 opened the pull request:
Use a more advanced strategy when picking out which working copy to remove, since sometimes the provided name of the file is only the last of the path.

Fixes #884

mickaelistria commented on org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacBindingResolver.java:
```diff
@@ -108,6 +108,7 @@ public class JavacBindingResolver extends BindingResolver {
 	public final IJavaProject javaProject;
 	private JavacConverter converter;
 	boolean isRecoveringBindings = false;
+	boolean hasResolved = false;
```
resolution may already have happened in JavacCompilationUnitResolver independently of bindings. So this flag can be false when resolution already happened.
I'm more curious about an example of why the `alreadyAnalyzed` condition can be false. Debugging the example could give a hint. Maybe the domToJavac is simply not well populated?

datho7561 replied:
I'm guessing at least one but not all of the symbols are populated, which prevents analyze from ever being called here. The root cause of the issue Gayan pinged me on was that some AST nodes (all the symbols in the class I was working with) didn't have symbols or types and I was expecting them to.

datho7561 replied:
I could switch it to check if all the javac nodes have symbols. I did made that change and ran the test suite, and it seems that sometimes its the case that all the javac nodes have symbols (because analysis has been run) but `this.symbolToDeclaration` hasn't been filled out yet.

mickaelistria commented on org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacBindingResolver.java:
```diff
@@ -434,9 +434,7 @@ private void resolve() {
-			boolean alreadyAnalyzed = this.converter.domToJavac.values().stream().map(TreeInfo::symbolFor).anyMatch(Objects::nonNull);
-			if (!alreadyAnalyzed) {
-				// symbols not already present: analyze
+			if (!this.converter.domToJavac.values().stream().map(TreeInfo::symbolFor).allMatch(Objects::nonNull)) {
```
Do we have the guarantee that all DCTree have a matching symbol? For example, what about blocks or for/if/while....? I don't think this condition is viable. It may work because it recomputes all bindings greedily, but it might not be fine enough.
Moreover, if the symptom is that `converter.domToJavac` sometimes miss values at that stage, we should look at the root cause of why some content is missing here. It can be a synchronization issue, or a conversion issue, or ...?

datho7561 replied:
Do you have some case for which https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/blob/969bc5a07fb352fce857a4f954cdbdc6e8f7ec0f/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacCompilationUnitResolver.java#L640 is used? I'm not sure when it happens. I set a breakpoint there and I haven't been able to get it to hit with the AST conversion or completion test suites, nor when playing around in the IDE.

mickaelistria replied:
This is only used when annotation processing is on (eg mapstruct or lombok or similar ones involved)

datho7561 replied:
Okay, thanks.

datho7561 replied:
> we should look at the root cause of why some content is missing here

I've been looking into this today. I identified one way to hack away the problem and pushed the code for that. The root cause is that one of the compilation units we process only has the last segment of the filepath. I'll need to keep looking to find the true root cause.

datho7561 replied:
Turns out the "last segment of the filepath" is the behaviour upstream as well. I have included a hack that should be able to prevent multiple versions of the same file from being included in the same compilation run, even when multiple files with the same name are open.

datho7561 committed:
Prevent error when renaming variable

Use a more advanced strategy when picking out which working copy to remove,
since sometimes the provided name of the file is only the last segment
of the path.

Fixes #884

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria commented on org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacCompilationUnitResolver.java:
```diff
@@ -503,9 +502,30 @@ public CompilationUnit toCompilationUnit(org.eclipse.jdt.internal.compiler.env.I
-		pathToUnit.put(new String(sourceUnit.getFileName()), sourceUnit);
+		
+		if (!pathToUnit.keySet().isEmpty() && sourceUnit instanceof BasicCompilationUnit) {
+			// BasicCompilationUnit sometimes just has the last segment of the path,
```
Should we also check that in the condition, eg add some `Path.of(fileName),getNameCount() == 1`? Could such a check replace the `instanceof BasicCompilationUnit` and thus make the code more self-explanatory?

mickaelistria commented on org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacCompilationUnitResolver.java:
```diff
@@ -503,9 +502,30 @@ public CompilationUnit toCompilationUnit(org.eclipse.jdt.internal.compiler.env.I
+			if (!realPaths.isEmpty()) {
+				// HACK: if the user has multiple files with the same name opened,
+				// remove all of them then add the one under analysis back
+				if (realPaths.size() > 1) {
```
Why not just overriding realPaths.get(0)` is it's the one we're replacing? Why do other units need to be discarded?

mickaelistria replied:
OK, I think I got it. Can you please then just add to comment "remove all of them as one of them could be a duplicate of the one under analysis"?

mickaelistria reviewed (commented):
I'm glad to see a finer grain fix.
I've just added questions/suggestions inline to hopefully improve the code.

datho7561 commented:
I'm going to wait to ensure there are no regressions; I don't want to have missed a null check somewhere
